File: go/linked-list/linked_list.go
package linkedlist

import (
	"errors"
)

type Node struct {
	Value    interface{}
	next     *Node
	previous *Node
}

type List struct {
	first *Node
	last  *Node
	size  int
}

var ErrEmptyList = errors.New("list has length zero")

func NewList(args ...interface{}) *List {
	newList := &List{}
	for _, value := range args {
		newList.Push(value)
	}
	return newList
}

func (n *Node) Next() *Node {
	return n.next
}

func (n *Node) Prev() *Node {
	return n.previous
}

func (l *List) Unshift(v interface{}) {
	l.first = &Node{Value: v, next: l.first}
	l.size++
	if l.size == 1 {
		l.last = l.first
	} else {
		l.first.next.previous = l.first
	}
}

func (l *List) Push(v interface{}) {
	l.last = &Node{Value: v, previous: l.last}
	l.size++
	if l.size == 1 {
		l.first = l.last
	} else {
		l.last.previous.next = l.last
	}
}

func (l *List) Shift() (interface{}, error) {
	if l.size == 0 {
		return nil, ErrEmptyList
	}
	value := l.first.Value
	if l.size == 1 {
		l.first, l.last = nil, nil
	} else {
		l.first = l.first.next
		l.first.previous = nil
	}
	l.size--
	return value, nil
}

func (l *List) Pop() (interface{}, error) {
	if l.size == 0 {
		return nil, ErrEmptyList
	}
	value := l.last.Value
	if l.size == 1 {
		l.first, l.last = nil, nil
	} else {
		l.last = l.last.previous
		l.last.next = nil
	}
	l.size--
	return value, nil
}

func (l *List) Reverse() {
	if l.size == 0 {
		return
	}
	l.last, l.first = l.first, l.last
	current := l.first
	for current != nil {
		current.next, current.previous = current.previous, current.next
		current = current.next
	}
}

func (l *List) First() *Node {
	return l.first
}

func (l *List) Last() *Node {
	return l.last
}

File: javascript/linked-list/linked-list.js
export class LinkedList {
  push(value) {
    let newObject = {value: value};
    if (!this._first) {
      [this._first, this._last] = [newObject, newObject];
    } else {
      newObject.previous = this._last;
      this._last.next = newObject;
      this._last = newObject;
    };
  }

  pop() {
    let value = this._last.value;
    if (!this._last.previous) {
      delete this._first, this._last;
    } else {
      this._last = this._last.previous;
      delete this._last.next;
    };
    return value;
  }

  shift() {
    let value = this._first.value;
    if (!this._first.next) {
      delete this._first, this._last;
    } else {
      this._first = this._first.next;
      delete this._first.previous;
    };
    return value;
  }

  unshift(value) {
    let newObject = {value: value};
    if (!this._first) {
      [this._first, this._last] = [newObject, newObject];
    } else {
      newObject.next = this._first;
      this._first.previous = newObject;
      this._first = newObject;
    };
  }

  delete(value) {
    if (!this._first) {
      return;
    } else {
      let currentObject = this._first;
      while (true) {
        if (currentObject.value === value) {
          if (currentObject.previous && currentObject.next) {
            currentObject.previous.next = currentObject.next;
            currentObject.next.previous = currentObject.previous;
          } else if (currentObject.previous) {
            this._last = currentObject.previous;
            delete this._last.next;
          } else if (currentObject.next) {
            this._first = currentObject.next;
            delete this._first.previous;
          } else {
            delete this._first, this._last;
          };
          break
        } else if (!currentObject.next) {
          break;
        } else {
          currentObject = currentObject.next;
        };
      };
    };
  }

  count() {
    if (!this._first) {
      return 0;
    } else {
      let [count, currentObject] = [1, this._first];
      while (currentObject.next) {
        count++;
        currentObject = currentObject.next;
      };
      return count;
    };
  }
}
